File: Groot priem getal generator/priemgetal.py
from random import randint
import secrets

def naar_binair(getal):
    binairGetal = []

    if getal == 0:
        binairGetal.insert(0, 0)
    else:
        while getal != 0:
            bit = getal % 2  # Modulus 2 om de rest van het getal / 2 te krijgen
            getal = getal // 2  # Getal delen door 2 voor volgende iteratie
            binairGetal.insert(0, bit)  # Bit (rest) opslagen in een lijst
    return binairGetal

def macht_mod(gndt, exp, mod):
    binair = naar_binair(exp)
    aantal_kwadraten = len(binair)
    resultaat = 1
    basis = gndt % mod
    for i in range(aantal_kwadraten - 1, -1, -1):
        if binair[i] == 1:
            resultaat = (resultaat * basis) % mod

        basis = (basis * basis) % mod
    
    return resultaat % mod

def willekeurig_1024_bit_getal():
    willekeurig_getal = secrets.randbits(1024)
    willekeurig_getal |= (1 << 1023) #Zorgen dat meest belangrijke bit 1 is zodat het getal uit 1024 bits bestaat
    willekeurig_getal |= 1 #Zorgen dat minst belangrijke bit 1 is zodat het getal oneven is
    return willekeurig_getal

def miller_rabin_test(getal, iterations=40):
    if getal in [2, 3]: #2 en 3 zijn priem
        return True
    if getal % 2 == 0: #Even getallen buiten 2 zijn niet priem
        return False
    n = getal
    n_1 = (getal - 1)
    k = 0
    while (n_1 % 2 == 0):
        n_1 = n_1 // 2
        k += 1
    m = n_1
    for _ in range(iterations):
        a = randint(2, (n - 2))
        b = macht_mod(a, m, n)
        if b != 1 and b != (n - 1):
            for _ in range(k - 1):
                b = (b * b) % n
                if b == (n - 1):
                    break
            else:
                return False #Samengesteld getal
    return True #Priem getal

def willekeurig_priem_1024_bit_getal():
    getal = willekeurig_1024_bit_getal()
    isPriem = miller_rabin_test(getal)
    while(isPriem == False):
        getal += 2
        isPriem = miller_rabin_test(getal)
    return getal

print(willekeurig_priem_1024_bit_getal())

File: Groot priem getal generator/priemgetal.cpp
#include <iostream>
#include <string>
#include <list>
#include <vector>
#include <random>
using namespace std;

list<int> naar_binair(int nummer) {
    list<int> binairGetal{};
    
    if(nummer == 0) {
        binairGetal.push_front(0);
    } else {
        while(nummer != 0) {
            int bit = nummer % 2; //Modulus 2 om de rest van het getal / 2 te krijgen
            nummer = nummer / 2; //Getal delen door 2 voor volgende iteratie
            binairGetal.push_front(bit); //Bit (rest) opslagen in een lijst
        }
    }

    return binairGetal;
}

int* list_naar_array(const list<int>& lst, size_t& elementen) {
    elementen = lst.size();
    int* arr = new int[elementen];
    auto it = lst.begin();
    for (size_t i = 0; i < elementen; ++i, ++it) {
        arr[i] = *it;
    }
    return arr;
}

int macht_mod(int gndt, int exp, int mod) {
    list<int> binair = naar_binair(exp);
    size_t aantal_kwadraten = binair.size();
    int* arrBinair = list_naar_array(binair, aantal_kwadraten);

    int resultaat = 1;
    int huidige_wortel = gndt % mod;

    for (size_t i = 0; i < aantal_kwadraten; ++i) {
        if (arrBinair[aantal_kwadraten - i - 1] == 1) {
            resultaat = (resultaat * huidige_wortel) % mod;
        }
        huidige_wortel = (huidige_wortel * huidige_wortel) % mod;
    }

    delete[] arrBinair;

    return resultaat;
}

vector<uint32_t> genereerRandom1024BitGetal() {
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<uint32_t> dis;

    vector<uint32_t> getal(32);
    for(size_t i = 0; i < getal.size(); ++i) {
        getal[i] = dis(gen);
    }

    getal.back() |= (uint32_t(1) << 31);

    return getal;
}

int main() {
    vector<uint32_t> random_number = genereerRandom1024BitGetal();

    // Output the generated number in hexadecimal
    for(size_t i = 0; i < random_number.size(); ++i) {
        printf("%08x", random_number[i]);
    }
}